joeynmt/training.py

```diff
@@ -213,44 +219,72 @@ def _save_checkpoint(self) -> None:
-            "optimizer_state": self.optimizer.state_dict(),
-            "scheduler_state": self.scheduler.state_dict() if
-            self.scheduler is not None else None,
-            'amp_state': amp.state_dict() if self.fp16 else None
+            "steps":
+                self.stats.steps,
+            "total_tokens":
+                self.stats.total_tokens,
+            "best_ckpt_score":
+                self.stats.best_ckpt_score,
+            "best_ckpt_iteration":
+                self.stats.best_ckpt_iter,
+            "model_state":
+                model_state_dict,
+            "optimizer_state":
+                self.optimizer.state_dict(),
+            "scheduler_state":
+                self.scheduler.state_dict()
+                if self.scheduler is not None else None,
+            'amp_state':
+                amp.state_dict() if self.fp16 else None,
+            "train_iter_state":
+                self.train_iter.state_dict()
         }
         torch.save(state, model_path)
-        if self.ckpt_queue.full():
-            to_delete = self.ckpt_queue.get()  # delete oldest ckpt
+        symlink_target = "{}.ckpt".format(self.stats.steps)
+        if new_best:
+            if len(self.ckpt_queue) == self.ckpt_queue.maxlen:
+                to_delete = self.ckpt_queue.popleft()  # delete oldest ckpt
+                try:
+                    os.remove(to_delete)
+                except FileNotFoundError:
+                    logger.warning(
+                        "Wanted to delete old checkpoint %s but "
+                        "file does not exist.", to_delete)
+
+            self.ckpt_queue.append(model_path)
+
+            best_path = "{}/best.ckpt".format(self.model_dir)
             try:
-                os.remove(to_delete)
-            except FileNotFoundError:
-                logger.warning("Wanted to delete old checkpoint %s but "
-                               "file does not exist.", to_delete)
-
-        self.ckpt_queue.put(model_path)
-
-        best_path = "{}/best.ckpt".format(self.model_dir)
-        try:
-            # create/modify symbolic link for best checkpoint
-            symlink_update("{}.ckpt".format(self.stats.steps), best_path)
-        except OSError:
-            # overwrite best.ckpt
-            torch.save(state, best_path)
-
-    def init_from_checkpoint(self, path: str,
+                # create/modify symbolic link for best checkpoint
+                symlink_update(symlink_target, best_path)
+            except OSError:
+                # overwrite best.ckpt
+                torch.save(state, best_path)
+
+        if self.save_latest_checkpoint:
+            last_path = "{}/latest.ckpt".format(self.model_dir)
+            previous_path = latest_checkpoint_update(symlink_target, last_path)
+            # If the last ckpt is in the ckpt_queue, we don't want to delete it.
+            can_delete = True
+            for ckpt_path in self.ckpt_queue:
```

Comment:
Shouldn't this check for `previous_checkpoint` being in the queue as well?

Reply:
I think I need to re-word that comment, it checks for `previous_path`, which is the path to the previous checkpoint. The latest checkpoint isn't being searched for in the queue.

Reply:
I see